README.md
# guiding-you-afteropen
guiding you source code after open


docs/CHANGELOG.txt
2007-2015 PrestaShop 

NOTICE OF LICENSE

This source file is subject to the Open Software License (OSL 3.0)
that is bundled with this package in the file LICENSE.txt.
It is also available through the world-wide-web at this URL:
http://opensource.org/licenses/osl-3.0.php
If you did not receive a copy of the license and are unable to
obtain it through the world-wide-web, please send an email
to license@prestashop.com so we can send you a copy immediately.

DISCLAIMER

Do not edit or add to this file if you wish to upgrade PrestaShop to newer
versions in the future. If you wish to customize PrestaShop for your
needs please refer to http://www.prestashop.com for more information.

@author PrestaShop SA <contact@prestashop.com>
@copyright 2007-2015 PrestaShop  SA
@license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
International Registred Trademark & Property of PrestaShop SA

Release Notes for PrestaShop 1.6
--------------------------------
####################################
#   v1.6.0.11 - (2015-01-07)
####################################

	Fixed bugs:

	[-] INSTALLER : Install default data on all languages installed if available && allow to install all available languages with CLI installer
	[-] INSTALLER : Translate the 2 recently added order states to Hebrew
	[-] INSTALLER : don't launch sessions on smarty compilation to avoid race conditions

	[-] FO : Fix bug #PSCSX-4211 wrong invoice by country.
	[-] FO : Could not see availability_statut when no available later value
	[-] FO : Fix bug #SCSX-3986, could not use blockmanufacturer select on index
	[-] FO : Fix bug #PSCSX-3601, invoice post code validation, thanks @Edwin
	[-] FO : fix the search when the order_by is configured as date_add
	[-] FO : Fix css bug #PSCSX-3842 payment divs classes
	[-] FO : Set the tax calculation method on PaymentModule::validateOrder() after loading the customer #PNM-2710
	[-] FO : Fix css bug PSCSX-3759
	[-] FO : Correct bad ajax cart behavior on mobile #PSCSX-3772
	[-] FO : Fix bug #PSCSX-3751, enable ssl on add to cart link in product link
	[-] FO : Fix bug #PSCSX-3771, bad update of summary on step 3
	[-] FO : Fix conversion rate to 1 if this is the default currency
	[-] FO : Don't write cookie when redirecting to an another protocol
	[-] FO : https detection using HTTP_X_FORWARDED_PROTO (use by cloudflare)
	[-] FO : Correct topmenu scrolling issue on touch devices #PNM-2943
	[-] FO : Fix bug #PSCSX-3162, no condition on virtual products
	[-] FO : Fix bug #PSCSX-3684 Bad prefix
	[-] FO : Stores : warning thrown if hours is empty
	[-] FO : FixBug #PSCSX-3652 - Wrong summary cart percent
	[-] FO : Product::getNewProducts() perfs improvements
	[-] FO : Fix wrapping fees tax excluded
	[-] FO : Fix bug #PSCSX-3549, bad back url when coming from fancybox
	[-] FO : Fix #PSCSX-3542, Convert currency in volume discount table (product.js)
	[-] FO : Fix bug #PSCSX-3500, bad post management for guest cehckout invoice
	[-] FO : Fix bug #PSCSX-3500, dni in guest + 5 steps
	[-] FO : Could not set lang by url
	[-] FO : Fix PSCFV-11878 fix update loyalty
	[-] FO : Fatal error on payment
	[-] FO : Fix bug #PSCFV-12537, bad minimal quantity for combinations
	[-] FO : fix bug #PSCSX-3418, bad colors list ordering in category
	[-] FO : Unicode support in minify regular expression
	[-] FO : Fix bug #PSCSX-3385, pagination translation in blocklayered
	[-] FO : Fix bug #PSCSX-2542, adding product to compare
	[-] FO : Fix bug #PSCSX-3201, could not sort DESC
	[-] FO : Fix bug #PSCSX-3304, keep customistion form url when ipa in url
	[-] FO : Fix bug #PSCSX-3304, jqzoom bad image on loading when ipa in url
	[-] FO : Fix PSCSX-3357 second part isCached bad call
	[-] FO : Fix PSCSX-3357 part fix add media
	[-] FO : Fix discount display issues #PSCSX-3308
	[-] FO : Fix bug #PSCSX-3304, could not retrieve ipa after saving customizations
	[-] FO : Fix bug #PSCSX-3281, display "doesntExist" warning even without stock management
	[-] FO : Fix #PSCSX-3211 no more duplicate product when reorder with cart_rule
	[-] FO : Fix bug #PSCSX-3243, blockcart hidden in colums on small screen
	[-] FO : Fix bug #PSCFV-12462, bad legend group by
	[-] FO : Fix PSCSX-3137 display thumbnail better
	[-] FO : Fix bug #PSCFV-12465, AuthController $back param
	[-] FO : Fix bug #PSCSX-3135, bad post cache on authentication.js
	[-] FO : Fix PSCSX-3090 fix custom product with minimum quantity
	[-] FO : Fix uniform mobile issue #PSCSX-1632
	[-] FO : Fix PSCSX-3062 display carrier when virtual + physical product in the same cart
	[-] FO : Fix the value of the wholesale price with combinations on order_detail table #PSCFV-8506
	[-] FO : Fix PSCSX-623 width in function of nb column diplayed
	[-] FO : FixBug #PSCSX-2893 - Wrong rowspan in shopping cart summary
	[-] FO : Fix bug #PSCSX-2991, 404 on cat image
	[-] FO : Fix PSCSX-2836 good redirect if multishipping
	[-] FO : Fix PSCSX-255 carrier not avaible if too heavy
	[-] FO : Fix bug #PSCSX-2972, basePriceDisplay.toFixed is not a function
	[-] FO : Fix PSCSX-2535 add asterisks for required fields
	[-] FO : Fix bug #PSCSX-2947, qty input when allowBuyWhenOutOfStock
	[-] FO : Fix PSCSX-2885 fix css width in blocklayered
	[-] FO : Fix css bug #PSCSX-2935, reduction on same line
	[-] FO : search combination upc, ean13 too
	[-] FO : added month names in identity.tpl so that they appear in the translations
	[-] FO : Fix PSCSX-2564 default set when deleting attributes default
	[-] FO : Fix PSCSX-2877 bad entity encoding in tinyMCE
	[-] FO : Fix bug #PSCSX-2651, debug smarty
	[-] FO : Correct responsive issue on category images #PSCSX-2843
	[-] FO : Fix #PSCSX-2484, Add style for .available-dif class
	[-] FO : extend mobile reponsive behavior to tablet #PSCSX-2842
	[-] FO : Correct left column responsive issue #PSCSX-1862

classes/CMS.php
<?php
/*
* 2007-2015 PrestaShop
*
* NOTICE OF LICENSE
*
* This source file is subject to the Open Software License (OSL 3.0)
* that is bundled with this package in the file LICENSE.txt.
* It is also available through the world-wide-web at this URL:
* http://opensource.org/licenses/osl-3.0.php
* If you did not receive a copy of the license and are unable to
* obtain it through the world-wide-web, please send an email
* to license@prestashop.com so we can send you a copy immediately.
*
* DISCLAIMER
*
* Do not edit or add to this file if you wish to upgrade PrestaShop to newer
* versions in the future. If you wish to customize PrestaShop for your
* needs please refer to http://www.prestashop.com for more information.
*
*  @author PrestaShop SA <contact@prestashop.com>
*  @copyright  2007-2015 PrestaShop SA
*  @license    http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
*  International Registered Trademark & Property of PrestaShop SA
*/

class CMSCore extends ObjectModel
{
	/** @var string Name */
	public $meta_title;
	public $meta_description;
	public $meta_keywords;
	public $content;
	public $link_rewrite;
	public $id_cms_category;
	public $position;
	public $indexation;
	public $active;

	/**
	 * @see ObjectModel::$definition
	 */
	public static $definition = array(
		'table' => 'cms',
		'primary' => 'id_cms',
		'multilang' => true,
		'fields' => array(
			'id_cms_category' => 	array('type' => self::TYPE_INT, 'validate' => 'isUnsignedInt'),
			'position' => 			array('type' => self::TYPE_INT),
			'indexation' =>     	array('type' => self::TYPE_BOOL),
			'active' => 			array('type' => self::TYPE_BOOL),

			// Lang fields
			'meta_description' => 	array('type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isGenericName', 'size' => 255),
			'meta_keywords' => 		array('type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isGenericName', 'size' => 255),
			'meta_title' =>			array('type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isGenericName', 'required' => true, 'size' => 128),
			'link_rewrite' => 		array('type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isLinkRewrite', 'required' => true, 'size' => 128),
			'content' => 			array('type' => self::TYPE_HTML, 'lang' => true, 'validate' => 'isCleanHtml', 'size' => 3999999999999),
		),
	);

	protected	$webserviceParameters = array(
		'objectNodeName' => 'content',
		'objectsNodeName' => 'content_management_system',
	);

	public function add($autodate = true, $null_values = false)
	{
		$this->position = CMS::getLastPosition((int)$this->id_cms_category);
		return parent::add($autodate, true);
	}

	public function update($null_values = false)
	{
		if (parent::update($null_values))
			return $this->cleanPositions($this->id_cms_category);
		return false;
	}

	public function delete()
	{
	 	if (parent::delete())
			return $this->cleanPositions($this->id_cms_category);
		return false;
	}

	public static function getLinks($id_lang, $selection = null, $active = true, Link $link = null)
	{
		if (!$link)
			$link = Context::getContext()->link;
		$result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS('
		SELECT c.id_cms, cl.link_rewrite, cl.meta_title
		FROM '._DB_PREFIX_.'cms c
		LEFT JOIN '._DB_PREFIX_.'cms_lang cl ON (c.id_cms = cl.id_cms AND cl.id_lang = '.(int)$id_lang.')
		'.Shop::addSqlAssociation('cms', 'c').'
		WHERE 1
		'.(($selection !== null) ? ' AND c.id_cms IN ('.implode(',', array_map('intval', $selection)).')' : '').
		($active ? ' AND c.`active` = 1 ' : '').
		'GROUP BY c.id_cms
		ORDER BY c.`position`');

		$links = array();
		if ($result)
			foreach ($result as $row)
			{
				$row['link'] = $link->getCMSLink((int)$row['id_cms'], $row['link_rewrite']);
				$links[] = $row;
			}
		return $links;
	}

	public static function listCms($id_lang = null, $id_block = false, $active = true)
	{
		if (empty($id_lang))
			$id_lang = (int)Configuration::get('PS_LANG_DEFAULT');

		return Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS('
		SELECT c.id_cms, l.meta_title
		FROM  '._DB_PREFIX_.'cms c
		JOIN '._DB_PREFIX_.'cms_lang l ON (c.id_cms = l.id_cms)
		'.Shop::addSqlAssociation('cms', 'c').'
		'.(($id_block) ? 'JOIN '._DB_PREFIX_.'block_cms b ON (c.id_cms = b.id_cms)' : '').'
		WHERE l.id_lang = '.(int)$id_lang.(($id_block) ? ' AND b.id_block = '.(int)$id_block : '').($active ? ' AND c.`active` = 1 ' : '').'
		GROUP BY c.id_cms
		ORDER BY c.`position`');
	}

	public function updatePosition($way, $position)
	{
		if (!$res = Db::getInstance()->executeS('
			SELECT cp.`id_cms`, cp.`position`, cp.`id_cms_category`
			FROM `'._DB_PREFIX_.'cms` cp
			WHERE cp.`id_cms_category` = '.(int)$this->id_cms_category.'
			ORDER BY cp.`position` ASC'
		))
			return false;

		foreach ($res as $cms)
			if ((int)$cms['id_cms'] == (int)$this->id)
				$moved_cms = $cms;

		if (!isset($moved_cms) || !isset($position))
			return false;

		// < and > statements rather than BETWEEN operator
		// since BETWEEN is treated differently according to databases
		return (Db::getInstance()->execute('
			UPDATE `'._DB_PREFIX_.'cms`
			SET `position`= `position` '.($way ? '- 1' : '+ 1').'
			WHERE `position`
			'.($way
				? '> '.(int)$moved_cms['position'].' AND `position` <= '.(int)$position
				: '< '.(int)$moved_cms['position'].' AND `position` >= '.(int)$position).'
			AND `id_cms_category`='.(int)$moved_cms['id_cms_category'])
		&& Db::getInstance()->execute('
			UPDATE `'._DB_PREFIX_.'cms`
			SET `position` = '.(int)$position.'
			WHERE `id_cms` = '.(int)$moved_cms['id_cms'].'
			AND `id_cms_category`='.(int)$moved_cms['id_cms_category']));
	}

	public static function cleanPositions($id_category)
	{
		$sql = '
		SELECT `id_cms`
		FROM `'._DB_PREFIX_.'cms`
		WHERE `id_cms_category` = '.(int)$id_category.'
		ORDER BY `position`';

		$result = Db::getInstance()->executeS($sql);

		for ($i = 0, $total = count($result); $i < $total; ++$i)
		{
			$sql = 'UPDATE `'._DB_PREFIX_.'cms`
					SET `position` = '.(int)$i.'
					WHERE `id_cms_category` = '.(int)$id_category.'
						AND `id_cms` = '.(int)$result[$i]['id_cms'];
			Db::getInstance()->execute($sql);
		}
		return true;
	}

	public static function getLastPosition($id_category)
	{
		$sql = '
		SELECT MAX(position) + 1
		FROM `'._DB_PREFIX_.'cms`
		WHERE `id_cms_category` = '.(int)$id_category;

		return (Db::getInstance()->getValue($sql));
	}

	public static function getCMSPages($id_lang = null, $id_cms_category = null, $active = true, $id_shop = null)
	{
		$sql = new DbQuery();
		$sql->select('*');
		$sql->from('cms', 'c');
		if ($id_lang)
			$sql->innerJoin('cms_lang', 'l', 'c.id_cms = l.id_cms AND l.id_lang = '.(int)$id_lang);

		if ($id_shop)
			$sql->innerJoin('cms_shop', 'cs', 'c.id_cms = cs.id_cms AND cs.id_shop = '.(int)$id_shop); 

		if ($active)
			$sql->where('c.active = 1');

		if ($id_cms_category)
			$sql->where('c.id_cms_category = '.(int)$id_cms_category);

		$sql->orderBy('position');

		return Db::getInstance()->executeS($sql);
	}

	public static function getUrlRewriteInformations($id_cms)
	{
	    $sql = 'SELECT l.`id_lang`, c.`link_rewrite`
				FROM `'._DB_PREFIX_.'cms_lang` AS c
				LEFT JOIN  `'._DB_PREFIX_.'lang` AS l ON c.`id_lang` = l.`id_lang`
				WHERE c.`id_cms` = '.(int)$id_cms.'
				AND l.`active` = 1';

		return Db::getInstance()->executeS($sql);
	}
}
